File: abc262/d/main.py
from collections import defaultdict, deque
from itertools import combinations, permutations
from bisect import bisect, bisect_left, bisect_right
from sortedcontainers import SortedSet, SortedDict, SortedList
import math
import sys
# from icecream import ic
sys.setrecursionlimit(4100000)
# def error(*args, end="\n"): print("[stderr]", *args, end=end, file=sys.stderr)
MOD = 998244353
INF = float("inf")
MINF = -float("inf")

N = int(input())
# N, K = map(int, input().split())
A = list(map(int, input().split()))

dp = [[[[0]*(N+1) for _ in range(N+1)] for _ in range(N+1)] for _ in range(N+1)]

for i in range(N+1):
    dp[0][0][i][0] = 1

ans = 0
for i in range(1, N+1): # i個目まで見る
    for j in range(N+1): # j個選ぶ
        for k in range(1, N+1): # kで割る
            for l in range(k): # あまり
                dp[i][j][k][l] += dp[i-1][j][k][l]
                dp[i][j][k][l] %= MOD
                if j > 0:
                    dp[i][j][k][l] += dp[i-1][j-1][k][(l-A[i-1])%k]
                    dp[i][j][k][l] %= MOD

for i in range(0, N+1):
    ans += dp[N][i][i][0]
    ans %= MOD
# ic(dp)
print(ans)

File: abc262/d/main.cpp
#include <iostream>
#include <vector>
using namespace std;

const int MOD = 998244353;
const int INF = 1e9;

int main() {
    int N;
    cin >> N;
    vector<int> A(N);
    for (int i = 0; i < N; i++) {
        cin >> A[i];
    }

    vector<vector<vector<vector<int>>>> dp(
        N + 1, vector<vector<vector<int>>>(N + 1, vector<vector<int>>(N + 1, vector<int>(N + 1, 0))));
    
    for (int i = 0; i <= N; i++) {
        dp[0][0][i][0] = 1;
    }

    int ans = 0;
    for (int i = 1; i <= N; i++) { // i個目まで見る
        for (int j = 0; j <= N; j++) { // j個選ぶ
            for (int k = 1; k <= N; k++) { // kで割る
                for (int l = 0; l < k; l++) { // あまり
                    dp[i][j][k][l] = dp[i - 1][j][k][l];
                    dp[i][j][k][l] %= MOD;
                    if (j > 0) {
                        dp[i][j][k][l] += dp[i - 1][j - 1][k][(l - A[i - 1] % k + k) % k];
                        dp[i][j][k][l] %= MOD;
                    }
                }
            }
        }
    }

    for (int i = 0; i <= N; i++) {
        ans += dp[N][i][i][0];
        ans %= MOD;
    }

    cout << ans << endl;
    return 0;
}